README.md

```diff
@@ -185,3 +186,82 @@ make test_unit
 # Shannon E2E test only
 make test_e2e_shannon_relay
 ```
+
+## User Data
+
+By default, PATH does not associate user data with service requests.
+
+You may opt to enable user data config to unlock the ability to associate a user with a service request.
+
+This is required for:
+
+- User specified app settings
+- Metering and billing of service requests
+- Rate limiting of service requests by throughput and/or capacity
+
+### Updated Endpoint
+
+Enabling user data will modify the endpoint for service requests to require a user app ID identifier at the end of the URL path.
+
+For example:
+
+```bash
+http://eth-mainnet.localhost:3000/v1/{user_app_id}
+```
+
+The default endpoint of `/v1` will no longer function without a user app ID.
+
+### Database Configuration
+
+To enable user data, you must set up a Postgres database and populate the `.config.yaml` file's `user_data_config` field with the connection string.
+
+```yaml
+user_data_config:
+  db_connection_string: "postgres://user:password@localhost:5432/database"
+```
+
+An example Postgres Docker configuration is included in the [docker-compose.yml](./docker-compose.yml) file at the root of this repository. **However, this configuration is not recommended for production use.**
+
+### Database Schema
+
+[Base Schema SQL File](./db/driver/sqlc/schema.sql)
+
+```mermaid
```

Comment:
PTAL at my comments in `user/user_app.go`.

I'm trying to understand things like:

1. `UserApp` or `GatewayApp`?
2. Is the `id` in a `UserApp` a `UserId` or an `AppId`?
3. What is an `AccountId`? Is it the same as a `UserId`?
4. What's a `SecretKey`? Is it the `ApplicationPrivateKey` from a protocol perspective?

For all of these, we can hop into a discussion or feel free to rename it in case the questions I asked trigger some sort of thought process.

Reply:
I've updated the comments in `user_app.go` to address these questions, but I will also answer them here.

At a high level, I am attempting to create a structure that will be roughly compatible with the current structure we use for account management and billing (note that what we call `PortalApp`s here are renamed to `UserApp`s):

<img width="1007" alt="Screenshot 2024-09-10 at 08 44 36" src="https://github.com/user-attachments/assets/830f6753-b856-46fb-b90f-1e4f2740b640">

As you can see in the structure above, an Account may have multiple UserApps and also may have many Users (a team), with only one plan. Accounts are used for Plan and team management, while Apps store individual settings.

> * 1. `UserApp` or `GatewayApp`?
[See comment here](https://github.com/buildwithgrove/path/pull/15#discussion_r1751527366)
> * 2. Is the `id` in a `UserApp` a `UserId` or an `AppId`?
It's an app ID. It's what gets passed in the URL path, eg. `rpc.grove.city/v1/{USER_APP_ID}`
> * 3. What is an `AccountId`? Is it the same as a `UserId`?
Not the same as a user ID. One Account may have many users with different permissions and one Plan. 
> * What's a `SecretKey`? Is it the `ApplicationPrivateKey` from a protocol perspective?
No. A secret key is simply an optional authentication key. 

Might be worth renaming to `API key` or something similar to avoid confusion.

 It's passed in the headers as `Authentication: {SECRET_KEY}` (Note: perhaps we should allow it being passed as `Authentication: Bearer {SECRET_KEY}` as that is a common pattern.)

Hopefully this makes things clear @Olshansk . CC @adshmh as well for your thoughts.

Reply:
I left some comments in `user_app.go` w.r.t the scope of this PR (i.e. removing auth and allow list data).

Also, per this explanation and the other comment, this is where my head is at right now (see the high level pseudo-struct below).

I'm not trying to over complicate, but trying to understand if this design is the right approach from first principles or just heavily influenced by how things are designed in portal-middleware today.

```go

type UserAccount {
	AccountId
	PlanType
	Auth
}

type GatewayEndpoint struct {
	// A unique ID specific to this endpoint
	EndpointID EndpointID

	UserAccount
	RateLimiting
}

type RateLimiting struct {
	ThroughputLimiting ...
	CapacityLimiting ...
}

type Auth struct {
	....
}
```

Reply:
> I left some comments in `user_app.go` w.r.t the scope of this PR (i.e. removing auth and allow list data).
> 
> Also, per this explanation and the other comment, this is where my head is at right now (see the high level pseudo-struct below).
> 
> I'm not trying to over complicate, but trying to understand if this design is the right approach from first principles or just heavily influenced by how things are designed in portal-middleware today.
> 
> ```go
> type UserAccount {
> 	AccountId
> 	PlanType
> 	Auth
> }
> 
> type GatewayEndpoint struct {
> 	// A unique ID specific to this endpoint
> 	EndpointID EndpointID
> 
> 	UserAccount
> 	RateLimiting
> }
> 
> type RateLimiting struct {
> 	ThroughputLimiting ...
> 	CapacityLimiting ...
> }
> 
> type Auth struct {
> 	....
> }
> ```

This looks fine except that I would suggest we keep Auth at the GatewayEndpoint (FKA UserApp) level.

Reply:
I've updated the `UserApp` struct to `GatewayEndpoint`. I think it's an improvement for sure.